dask/dataframe/tests/test_dataframe.py

```diff
@@ -1426,13 +1442,92 @@ def test_concat2():
     b = dd.DataFrame(dsk, 'y', ['a', 'b'], [None, None])
 
     c = dd.concat([a, b])
+    assert eq(c, pd.concat([a.compute(), b.compute()]))
 
     assert c.npartitions == a.npartitions + b.npartitions
-
+    assert c.divisions == (None, ) * (c.npartitions + 1)
     assert eq(pd.concat([a.compute(), b.compute()]), c)
-
     assert dd.concat([a, b]).dask == dd.concat([a, b]).dask
 
+    dsk = {('x', 0): pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]}),
+           ('x', 1): pd.DataFrame({'a': [4, 5, 6], 'b': [3, 2, 1]})}
+    a = dd.DataFrame(dsk, 'x', ['a', 'b'], [None, None])
+    dsk = {('y', 0): pd.DataFrame({'b': [10, 20, 30], 'c': [40, 50, 60]}),
+           ('y', 1): pd.DataFrame({'b': [40, 50, 60], 'c': [30, 20, 10]})}
+    b = dd.DataFrame(dsk, 'y', ['b', 'c'], [None, None])
+
+    c = dd.concat([a, b])
+    assert eq(c, pd.concat([a.compute(), b.compute()]))
+
+
+def test_concat3():
+    pdf1 = pd.DataFrame(np.random.randn(6, 5),
+                        columns=list('ABCDE'), index=list('abcdef'))
+    pdf2 = pd.DataFrame(np.random.randn(6, 5),
+                        columns=list('ABCFG'), index=list('ghijkl'))
+    pdf3 = pd.DataFrame(np.random.randn(6, 5),
+                        columns=list('ABCHI'), index=list('mnopqr'))
+    ddf1 = dd.from_pandas(pdf1, 2)
+    ddf2 = dd.from_pandas(pdf2, 3)
+    ddf3 = dd.from_pandas(pdf3, 2)
+
+    result = dd.concat([ddf1, ddf2])
+    assert result.divisions == ddf1.divisions[:-1] + ddf2.divisions
+    assert result.npartitions == ddf1.npartitions + ddf2.npartitions
+    assert eq(result, pd.concat([pdf1, pdf2]))
+
+    assert eq(dd.concat([ddf1, ddf2], respect_partitions=True),
+              pd.concat([pdf1, pdf2]))
+
+    result = dd.concat([ddf1, ddf2, ddf3])
+    assert result.divisions == (ddf1.divisions[:-1] + ddf2.divisions[:-1] +
+                                ddf3.divisions)
+    assert result.npartitions == (ddf1.npartitions + ddf2.npartitions +
+                                  ddf3.npartitions)
+    assert eq(result, pd.concat([pdf1, pdf2, pdf3]))
+
+    assert eq(dd.concat([ddf1, ddf2, ddf3], respect_partitions=True),
+              pd.concat([pdf1, pdf2, pdf3]))
+
+def test_concat4_respect_partitions():
+    pdf1 = pd.DataFrame(np.random.randn(10, 5),
+                        columns=list('ABCDE'), index=list('abcdefghij'))
+    pdf2 = pd.DataFrame(np.random.randn(13, 5),
+                        columns=list('ABCDE'), index=list('fghijklmnopqr'))
+    pdf3 = pd.DataFrame(np.random.randn(13, 6),
+                        columns=list('CDEXYZ'), index=list('fghijklmnopqr'))
+
+    ddf1 = dd.from_pandas(pdf1, 2)
+    ddf2 = dd.from_pandas(pdf2, 3)
+    ddf3 = dd.from_pandas(pdf3, 2)
+
+    assert raises(ValueError, lambda: dd.concat([ddf1, ddf1]))
```

Comment:
Can we test that this raises an informative error pointing users to use `respect_partitions=True`?  E.g.

```python
try:
    dd.concat([ddf1, ddf1])
except ValueError as e:
    assert 'respect_partitions=True' in str(e)
```